File: backend/src/schemas/block.ts
import { date, integer, pgTable, serial, text, timestamp, unique } from 'drizzle-orm/pg-core'
import { hourTable } from './hour'
import { createInsertSchema, createSelectSchema } from 'drizzle-zod'

export const blockTable = pgTable('block', {
  id: serial('id').notNull().primaryKey(),
  hourId: integer('hour_id').notNull().references(() => hourTable.id),
  date: date('date').notNull(),
  period: text('period', { enum: ['matutine', 'vespertine'] }),
  createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow()
}, (table) => ({
  uniqueBlock: unique('unique_block').on(table.hourId, table.date)
}))

export const selectBlockSchema = createSelectSchema(blockTable)
export const insertBlockSchema = createInsertSchema(blockTable)

File: flutter_frontend/lib/classes/block.dart
class Block {
  final int id;
  final int hourId;
  final String date;
  final String period;
  final String createdAt;
  final String updatedAt;

  Block({
    required this.id,
    required this.hourId,
    required this.date,
    required this.period,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Block.fromJson(Map<String, dynamic> json) {
    return switch (json) {
      {
        "id": int id,
        "hourId": int hourId,
        "date": String date,
        "period": String period,
        "createdAt": String createdAt,
        "updatedAt": String updatedAt,
      } =>
        Block(
          id: id,
          hourId: hourId,
          date: date,
          period: period,
          createdAt: createdAt,
          updatedAt: updatedAt,
        ),
      _ => throw const FormatException("Block: wrong format")
    };
  }
}
